### PR TITLE
feat: initial HashJoinRel support

### PR DESCRIPTION
**Pull request by vibhatha**
This PR includes the base work for integrating HashEquiJoin.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ Release Notes
 
 * StrLiteral is no longer converted to CHAR(<length>)
 
-fix(isthmus): convert BinaryLiteral to VARBINARY 
+fix(isthmus): convert BinaryLiteral to VARBINARY
```
Comment (danepitkin) [inline CHANGELOG.md:11]: nit: maybe revert whitespace change?

File: core/src/main/java/io/substrait/relation/AbstractJoin.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package io.substrait.relation;
+
+import io.substrait.expression.Expression;
+import java.util.Optional;
+
+public abstract class AbstractJoin extends BiRel implements HasExtension {
+
+  public abstract Optional<Expression> getPostJoinFilter();
```
Comment (danepitkin) [inline core/src/main/java/io/substrait/relation/AbstractJoin.java:8]: I just want to add that a NestedLoopJoin does not have a post join filter, so it wouldn't be able to use this abstract class. Maybe it's not needed after all?

**Review by danepitkin — COMMENTED** [review]
Nice work!

File: CHANGELOG.md (hunk 1 above)
Comment (vibhatha) [reply CHANGELOG.md:11]: I got a formatting issue, that's why I added it, but let me check it again.

File: core/src/main/java/io/substrait/relation/AbstractJoin.java (hunk 2 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/AbstractJoin.java:8]: Ah good point. I will update the PR.

File: CHANGELOG.md (hunk 1 above)
Comment (vibhatha) [reply CHANGELOG.md:11]: @danepitkin I just went through NLJ PR to Substrait-Java, and I see something similar.

https://github.com/substrait-io/substrait-java/actions/runs/6475823135/job/17583520911#step:4:1

File: core/src/main/java/io/substrait/relation/physical/HashJoin.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  public abstract Optional<List<FieldReference>> getRightKeys();
+
+  public abstract JoinType getJoinType();
+
+  @Override
+  protected Type.Struct deriveRecordType() {
+    Stream<Type> leftTypes =
+        switch (getJoinType()) {
+          case RIGHT, OUTER -> getLeft().getRecordType().fields().stream()
```
Comment (danepitkin) [inline core/src/main/java/io/substrait/relation/physical/HashJoin.java:27]: Should we add RIGHT_SEMI and RIGHT_ANTI? (Same with LEFT below)

File: CHANGELOG.md (hunk 1 above)
Comment (danepitkin) [reply CHANGELOG.md:11]: ohhh, is that what my error is?? Good catch, thank you

Comment (vibhatha) [reply CHANGELOG.md:11]: It seems like that, but how could it be there in the first place?

File: core/src/main/java/io/substrait/relation/physical/HashJoin.java (hunk 3 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/physical/HashJoin.java]: Good catch. no columns from the opposite are added.

File: core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java [hunk 4]
```diff
@@ -174,6 +175,21 @@ void join() {
     verifyRoundTrip(rel);
   }
 
+  @Test
+  void hashJoin() {
+    int[] left_keys = {};
```
Comment (EpsilonPrime) [inline core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java:180]: Should we have some tests with non-empty keys to make sure that's working correctly?

File: CHANGELOG.md (hunk 1 above)
Comment (vbarua) [reply CHANGELOG.md:11]: This is extra weird because the changelog is generated programatically 🤔 

Ah, the whitespace is in the commit message itself
https://github.com/substrait-io/substrait-java/commit/6e82f397a168f3811593fe8248753d70d4f0c04f

I think it's fine to fix it like this.

File: core/src/main/java/io/substrait/relation/physical/HashJoin.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class HashJoin extends BiRel implements HasExtension {
+
+  public abstract Optional<List<FieldReference>> getLeftKeys();
+
+  public abstract Optional<List<FieldReference>> getRightKeys();
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/physical/HashJoin.java:21]: I would argue that both `getLeftKeys` and `getRightKeys` should be required fields for a HashJoin (i.e. `List<FieldReference>`) as users should be forced to set them. As is, a user can construct a HashJoin with only left keys or only right keys.

Admittedly they can do this in the proto as well as the fields are not marked as required, but we can choose to be stricter in the library to guide their hand. If a user really wants a hash join on no keys, they should set the keys to empty lists.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java [hunk 6]
```diff
@@ -228,6 +235,37 @@ public Rel visit(ExtensionTable extensionTable) throws RuntimeException {
...
+
+    hashJoin
+        .getRightKeys()
+        .ifPresent(
+            keys -> {
+              for (int i = 0; i < keys.size(); i++) {
+                builder.setRightKeys(i, toProto(keys.get(i)));
+              }
+            });
+
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/RelProtoConverter.java:264]: We should also set the post join condition

File: core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java [hunk 7]
```diff
@@ -166,6 +168,29 @@ public Optional<Rel> visit(Cross cross) throws RuntimeException {
...
+  @Override
+  public Optional<Rel> visit(HashJoin hashJoin) throws RuntimeException {
+    var left = hashJoin.getLeft().accept(this);
+    var right = hashJoin.getRight().accept(this);
+    var leftKeys = hashJoin.getLeftKeys();
+    var rightKeys = hashJoin.getRightKeys();
+    var postFilter = hashJoin.getPostJoinFilter().flatMap(t -> visitExpression(t));
+    if (allEmpty(left, right, leftKeys, rightKeys, postFilter)) {
+      return Optional.empty();
+    }
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java:180]: The purpose of the RelCopyOnWriteVisitor is to produce a copy of a plan tree ONLY if some part of the tree has changed.

Here, `left`, `right` and `postFilter` are all empty if they have not been changed by calling `accept` or `visitExpression`, so it makes sense to include them in the `allEmpty` check 

`leftKeys` and `rightKeys` just return optional if they are not set (possibly another reason to have them be required in HashJoin). As is, they shouldn't be in the allEmpty check.

There is a deeper issue in that field references possibly _should_ be recomputed as part of the RelCopyOnWriteVisitor (see https://github.com/substrait-io/substrait-java/issues/185), but I think that's outside the scope of the PR. Noting it here so it gets linked to the issue.

File: core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java (hunk 4 above)
Comment (vbarua) [reply core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java:180]: +1 to using non-empty keys.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 6 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:264]: Heh, the Join converter is also missing the post-join condition 😅

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 8]
```diff
@@ -165,6 +166,28 @@ private Join join(
         .build();
   }
 
+  public HashJoin hashJoin(
+      int[] leftKeys, int[] rightKeys, HashJoin.JoinType joinType, Rel left, Rel right) {
```
Comment (vbarua) [inline core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:170]: I would avoid using bare arrays here as it makes this hard to use in my opinion. With something like:
```java
  public HashJoin hashJoin(
      List<Integer> leftKeys, List<Integer> rightKeys, HashJoin.JoinType joinType, Rel left, Rel right)
```
users can inline the key lists like
```java
b.hashJoin(
  Collections.EMPTY_LIST,
  Collections.EMPTY_LIST,
  HashJoin.JoinType.INNER, commonTable, commonTable))
```
or
```java
b.hashJoin(
  List.of(1,3),
  List.of(2,4),
  HashJoin.JoinType.INNER, commonTable, commonTable)
```
assuming they have access to newer versions of Java which _we_ don't within the core library, but we do with the isthmus side of things.

**Review by vbarua — CHANGES_REQUESTED** [review]
Overall this looks really good. Left some comments.

File: core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java (hunk 7 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java:180]: Shall we create an issue? And make sure this is addressed.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java (hunk 8 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:170]: Ah I was thinking a better way 😃
I will update.

**Comment by vibhatha** [comment]
@vbarua Thank you for the review. I will address the comments.

File: core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java (hunk 7 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java:180]: I think https://github.com/substrait-io/substrait-java/issues/185 should be enough. It captures the overall issue.

File: core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java (hunk 4 above)
Comment (vibhatha) [reply core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java:180]: @vbarua @EpsilonPrime 
Looking into the `commonTable` initialization, we do have empty lists to create it, so if we use non-empty lists for keys, we may need to introduce a commonTable with non-empty initialization? Am I following this incorrectly?


```java
final Rel commonTable =
      b.namedScan(Collections.emptyList(), Collections.emptyList(), Collections.emptyList());
```

File: core/src/main/java/io/substrait/relation/physical/HashJoin.java (hunk 5 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/physical/HashJoin.java:21]: This is a good point. I think I didn't think about it well in the first go. +1 for this change. 
I will add it.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 6 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:264]: I will add it too.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java [hunk 9]
```diff
@@ -228,6 +237,33 @@ public Rel visit(ExtensionTable extensionTable) throws RuntimeException {
...
+    List<FieldReference> rightKeys = hashJoin.getRightKeys();
+
+    if (leftKeys.size() != rightKeys.size()) {
+      throw new RuntimeException("Number of left and right keys must be equal.");
+    }
+
+    for (int idx = 0; idx < hashJoin.getLeftKeys().size(); idx++) {
+      builder.setLeftKeys(idx, toProto(leftKeys.get(idx)));
+      builder.setRightKeys(idx, toProto(rightKeys.get(idx)));
+    }
```
Comment (vibhatha) [inline core/src/main/java/io/substrait/relation/RelProtoConverter.java:259]: @vbarua can I do better here? 
Also validation of left and right key sizes is a bit doubtful.

File: core/src/main/java/io/substrait/relation/physical/HashJoin.java (hunk 5 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/physical/HashJoin.java]: I updated the code.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 6 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:262]: Updated the code.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java (hunk 8 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java]: Updated the code.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 9 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:259]: The validation seems reasonable to me. The only thing I would suggest to improve it would be to provide additional context. Something like:
```
The number of left and right keys must be equal for HashJoin. Received left %s and right %s keys.
```

For setting the keys themselves, I think
```java
    builder.addAllLeftKeys(leftKeys.stream().map(this::toProto).collect(Collectors.toList()));
    builder.addAllRightKeys(rightKeys.stream().map(this::toProto).collect(Collectors.toList()));
```
is a bit cleaner.

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 6 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:262]: Thanks for fixing it in both places 🙇

File: core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java (hunk 4 above)
Comment (vbarua) [reply core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java]: > we may need to introduce a commonTable with non-empty initialization? 

This is correct. You can use the `b.namedScan` method to create a table with columns, something like
```java
b.namedScan(List.of("example"), List.of("a"), List.of(R.STRING)
```

One small structural thing. The tests in `ExtensionRoundtripTest.java` verify that _extensions_ roundtrip. The test that you've added already does this perfectly well.

For adding a test using non-empty keys, could I ask that you introduce something like `JoinRountripTest.java` (in a similar vein to `AggregateRoundtripTest.java`) and do it there. That will also be a good place to add nested join tests later.

File: core/src/main/java/io/substrait/relation/ProtoRelConverter.java [hunk 10]
```diff
@@ -490,6 +495,35 @@ private Set newSet(SetRel rel) {
...
+    Type.Struct unionedStruct = Type.Struct.builder().from(leftStruct).from(rightStruct).build();
+    var converter = new ProtoExpressionConverter(lookup, extensions, unionedStruct, this);
+    var builder =
+        HashJoin.builder()
+            .left(left)
+            .right(right)
+            .leftKeys(
+                rel.getLeftKeysList().stream().map(converter::from).collect(Collectors.toList()))
+            .rightKeys(
+                rel.getRightKeysList().stream().map(converter::from).collect(Collectors.toList()))
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/ProtoRelConverter.java:512]: I _think_ there might be a bug here. The offsets of the right key list are relative to the right input. However, here the converter is created using the left _and_ right inputs. To illustrate, given two tables like:
```
t1: (a: i64, b: fp64, c:string)
t2: (x: fp64, y: string, z: i64)
```
let's say we want to join on
```
left keys: (0, 1)
right keys: (2, 0)
```
the converter has the following root type
```
(a: i64, b: fp64, c:string, x: fp64, y: string, z: i64)
```
so the left keys `(0,1)` refer to the correct inputs `(a, b)` but the right keys `(2, 0)` refer to `(c, a)` not `(z, x)`.

It would be good to have a test case like this.

File: core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java [hunk 11]
```diff
@@ -166,6 +168,29 @@ public Optional<Rel> visit(Cross cross) throws RuntimeException {
...
+    return Optional.of(
+        ImmutableHashJoin.builder()
+            .from(hashJoin)
+            .left(left.orElse(hashJoin.getLeft()))
+            .right(right.orElse(hashJoin.getRight()))
+            .leftKeys(leftKeys)
+            .rightKeys(rightKeys)
+            .postJoinFilter(
+                Optional.ofNullable(
+                    postFilter.orElseGet(() -> hashJoin.getPostJoinFilter().orElse(null))))
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java:190]: Because postFilter is already an Optional, we can use the `or` method instead of `Optional.ofNullable` here like:
```java
            .postJoinFilter(postFilter.or(hashJoin::getPostJoinFilter))
```

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 9 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java:259]: This is perfect. I will update.

File: core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java (hunk 4 above)
Comment (vibhatha) [reply core/src/test/java/io/substrait/type/proto/ExtensionRoundtripTest.java]: @vbarua I will defintely add a test class as you have susggested here. 👍

File: core/src/main/java/io/substrait/relation/ProtoRelConverter.java (hunk 10 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/ProtoRelConverter.java]: @vbarua you are correct. Thanks for catching this. It did lack the non-key test and it uncovered this issue. 

I added a converter separately for left and right, and then kept the union-based one for the post join filter. 

Also there are redundant parts in the newly added join test class with respect to the extension roundrip tests. Shall we use another abstraction to keep the common utils?

File: core/src/main/java/io/substrait/relation/RelProtoConverter.java (hunk 9 above)
Comment (vibhatha) [reply core/src/main/java/io/substrait/relation/RelProtoConverter.java]: Updated.

**Comment by vibhatha** [comment]
@vbarua ready for another round of review and left a few questions too 🙂

File: core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java (hunk 11 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java:190]: You're not missing anything, I forget that we're on Java 8 sometimes so we don't get access to all the nice Optional util methods 🥲

File: core/src/main/java/io/substrait/relation/ProtoRelConverter.java [hunk 12]
```diff
@@ -490,6 +496,39 @@ private Set newSet(SetRel rel) {
     return builder.build();
   }
 
+  private Rel newHashJoin(HashJoinRel rel) {
+    Rel left = from(rel.getLeft());
+    Rel right = from(rel.getRight());
+    var leftKeys = rel.getLeftKeysList();
+    var rightKeys = rel.getRightKeysList();
+    var rightOffSetKeys =
+        Stream.concat(leftKeys.stream(), rightKeys.stream()).collect(Collectors.toList());
```
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/ProtoRelConverter.java:505]: `rightOffSetKeys` isn't used, could you remove it.

File: core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+      AdvancedExtension.builder()
+          .enhancement(new StringHolder("COMMON ENHANCEMENT"))
+          .optimization(new StringHolder("COMMON OPTIMIZATION"))
+          .build();
+
+  final AdvancedExtension relExtension =
+      AdvancedExtension.builder()
+          .enhancement(new StringHolder("REL ENHANCEMENT"))
+          .optimization(new StringHolder("REL OPTIMIZATION"))
+          .build();
```
Comment (vbarua) [inline core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java:56]: We don't need to build and test common and rel extensions here, the one test we have in ExtensionRoundtripTest verifies it.

File: core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  void verifyRoundTrip(Rel rel) {
+    io.substrait.proto.Rel protoRel = relProtoConverter.toProto(rel);
+    Rel relReturned = protoRelConverter.from(protoRel);
+    assertEquals(rel, relReturned);
+  }
+
+  @Test
+  void hashJoin() {
+    List<Integer> leftEmptyKeys = Arrays.asList(0, 1);
+    List<Integer> rightEmptyKeys = Arrays.asList(2, 0);
```
Comment (vbarua) [inline core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java:67]: minor: `leftEmptyKeys` -> `leftKeys` and `rightEmptyKeys` -> `rightKeys`

The keys are no longer empty 🙂

File: core/src/main/java/io/substrait/relation/ProtoRelConverter.java (hunk 10 above)
Comment (vbarua) [reply core/src/main/java/io/substrait/relation/ProtoRelConverter.java]: > Also there are redundant parts in the newly added join test class with respect to the extension roundrip tests. Shall we use another abstraction to keep the common utils?

I think it would be worth doing at some point, but I don't think it needs to be part of this PR. I'd rather keep the changeset small.

**Review by vbarua — COMMENTED** [review]
Overall change looks good. Left some final comments. I think we can get this in next week 🙂

File: core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java (hunk 14 above)
Comment (vibhatha) [reply core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java:67]: Oh 😲

**Comment by vibhatha** [comment]
@vbarua I will make these changes and thank you for the quick response 🙏

**Comment by vibhatha** [comment]
@vbarua addressed the reviews 🙂

**Review by vbarua — APPROVED** [review]
LGMT! Thanks for contribution 🙇

**Comment by vibhatha** [comment]
Appreciate your support :pray:
